File: prune_saenet/impute_by_mice.r
if (!require(mice)) {
    quit("no", 2)
}

library(mice)

# command args
temp_dir <- NULL
seed <- NULL

args <- commandArgs(trailingOnly=TRUE)
for (arg in args) {
    s <- strsplit(arg, "=")[[1]]
    if (length(s) < 2) {
        next
    }

    if (s[1] == "--temp_dir") {
        temp_dir <- paste(s[2:length(s)], collapse="=")
    } else if (s[1] == "--rs") {
        seed <- strtoi(s[2])
    }
}

if (!is.null(seed))
    set.seed(seed)

# args for mice
X <- NULL
n_imputations <- NULL
maxit <- NULL
is_discrete <- NULL

# X
path <- file.path(temp_dir, "X.csv")
X <- read.csv(path, sep=',', header=FALSE)

# n_imputations
path <- file.path(temp_dir, "n_imputations.csv")
n_imputations <- read.csv(path, sep=',', header=FALSE)[1, 1]

# maxit
path <- file.path(temp_dir, "maxit.csv")
maxit <- read.csv(path, sep=',', header=FALSE)[1, 1]

# is_discrete
path <- file.path(temp_dir, "is_discrete.csv")
if (file.exists(path)) {
    is_discrete <- read.csv(path, sep=",", header=FALSE)
    is_discrete <- as.vector(is_discrete)
}

# run mice
# XXX: 一回目は、離散を含んでもmethの指定はいらないのだろうか？
mids <- mice(X, m=n_imputations, method="norm", maxit=maxit)

meth <- mids$method
if (!is.null(is_discrete)) {
    for (i in 1:length(is_discrete)) {
        if (!is_discrete[[i]]) next
        meth[names(meth)[i]] <- "logreg"
    }
}

mids <- mice(X, m=n_imputations, method=meth, maxit=maxit)
X_imputed <- lapply(1:n_imputations, function(i) complete(mids, action = i))

# output
names <- c()
for (i in 1:n_imputations) {
    fname <- paste("X_imputed_", sprintf("%08d", i), ".csv", sep="")
    path <- file.path(temp_dir, fname)
    write.csv(X_imputed[i], path, row.names=FALSE)
    
    names <- c(names, fname)
}

path <- file.path(temp_dir, "result_filenames.csv")
write.csv(names, path, row.names=F, quote=F)

quit("no", 0)

File: prune_saenet/_impute_by_mice.py
import os
import shutil
import tempfile
import subprocess

import numpy as np
from sklearn.utils import check_array, check_scalar


def impute_by_mice(X, n_imputations=10, maxit=10, is_discrete=None, seed=None):
    """ multiple imputation by mice.

    Parameters
    ----------
    X : array-like, shape (n_samples, n_features)
        Target data, where ``n_samples`` is the number of samples
        and ``n_features`` is the number of features.
    n_imputations : int, optional (defualt=10)
        The number of multiple imputations.
    maxit : int, optional (default=10)
        The number of iterations.
    is_discrete : array-like of shape (n_features, ), optional (default=None)
        ``is_discrete[i]`` specifies whether the i-th feature is discrete or not.
        All features are considered continuous if ``is_discrete`` is None.
    seed : int, optional (default=None)
        The seed for random numbers.

    Returns
    -------
    imputed_X_list : list, shape [X, ...]
        The list containing imputed X.
    """

    # check arguments
    X = check_array(X, force_all_finite="allow-nan")

    n_imputations = check_scalar(n_imputations, "n_imputations", int)

    maxit = check_scalar(maxit, "maxit", int)

    if is_discrete is not None:
        is_discrete = check_array(is_discrete, dtype=bool, ensure_2d=False)

    if seed is not None:
        seed = check_scalar(seed, "seed", int)

    try:
        # args
        temp_dir = tempfile.mkdtemp()

        path = os.path.join(os.path.dirname(__file__), "impute_by_mice.r")

        args = [f"--temp_dir={temp_dir}"]
        if seed is not None:
            args.append(f"--rs={seed}")

        # write data
        np.savetxt(temp_dir + "/X.csv", X, delimiter=",", fmt="%s")

        # write params
        np.savetxt(temp_dir + "/n_imputations.csv", [n_imputations], delimiter=",", fmt="%s")
        np.savetxt(temp_dir + "/maxit.csv", [maxit], delimiter=",", fmt="%d")
        if is_discrete is not None:
            is_discrete = [str(e).upper() for e in is_discrete]
            np.savetxt(temp_dir + "/is_discrete.csv", [is_discrete], delimiter=",", fmt="%s")

        # run
        ret = subprocess.run(["Rscript", path, *args], capture_output=True)
        if ret.returncode != 0:
            print(ret.stdout.decode())
            print(ret.stderr.decode())

        # retrieve imputed datasets
        imputed_X_list = []
        fnames = np.loadtxt(temp_dir + "/result_filenames.csv", delimiter=",", skiprows=1, dtype=str)
        for fname in fnames:
            imputed_X = np.loadtxt(temp_dir + "/" + fname, delimiter=",", skiprows=1)
            imputed_X_list.append(imputed_X)
    except FileNotFoundError:
        raise RuntimeError("Rscript is not found.")
    except BaseException as e:
        raise RuntimeError(str(e))
    finally:
        if os.path.exists(temp_dir):
            shutil.rmtree(temp_dir)

    return imputed_X_list
